biopet-utils/src/main/scala/nl/lumc/sasc/biopet/utils/summary/db/SummaryDb.scala

```diff
@@ -448,6 +346,138 @@ class SummaryDb(val db: Database) extends Closeable {
+    var q: Query[Executables, Executables#TableElementType, Seq] = executables
+    runId.foreach(r => q = q.filter(_.runId === r))
+    toolName.foreach(r => q = q.filter(_.toolName === r))
+    q
+  }
+
+  /** Return all executables with given criteria */
+  def getExecutables(runId: Option[Int] = None, toolName: Option[String] = None): Future[Seq[Executable]] = {
+    db.run(executablesFilter(runId, toolName).result)
+  }
+
+}
+
+class SummaryDbReadOnly(val db: Database) extends SummaryDb
+
+class SummaryDbWrite(val db: Database) extends SummaryDb {
+  /** This method will create all tables */
+  def createTables(): Unit = {
+    try {
+      val setup = DBIO.seq(
+        (runs.schema ++ samples.schema ++
+          libraries.schema ++ pipelines.schema ++
+          modules.schema ++ stats.schema ++ settings.schema ++
+          files.schema ++ executables.schema).create
+      )
+      val setupFuture = db.run(setup)
+      Await.result(setupFuture, Duration.Inf)
+    }
+  }
+
+  /** This method will create a new run and return the runId */
+  def createRun(runName: String, outputDir: String, version: String, commitHash: String,
+                creationDate: Date): Future[Int] = {
+    val id = Await.result(db.run(runs.size.result), Duration.Inf)
+    db.run(runs.forceInsert(Run(id, runName, outputDir, version, commitHash, creationDate))).map(_ => id)
+  }
+
+  /** This creates a new sample and return the sampleId */
+  def createSample(name: String, runId: Int, tags: Option[String] = None): Future[Int] = {
+    val id = Await.result(db.run(samples.size.result), Duration.Inf)
+    db.run(samples.forceInsert(Sample(id, name, runId, tags))).map(_ => id)
+  }
+
+  def createOrUpdateSample(name: String, runId: Int, tags: Option[String] = None): Future[Int] = {
+    getSampleId(runId, name).flatMap(_ match {
+      case Some(id: Int) =>
+        db.run(samples.filter(_.name === name).filter(_.id === id).map(_.tags).update(tags))
+          .map(_ => id)
+      case _ => createSample(name, runId, tags)
+    })
+  }
+
+  /** This will create a new library */
+  def createLibrary(name: String, runId: Int, sampleId: Int, tags: Option[String] = None): Future[Int] = {
+    val id = Await.result(db.run(libraries.size.result), Duration.Inf)
+    db.run(libraries.forceInsert(Library(id, name, runId, sampleId, tags))).map(_ => id)
+  }
+
+  def createOrUpdateLibrary(name: String, runId: Int, sampleId: Int, tags: Option[String] = None): Future[Int] = {
+    getLibraryId(runId, sampleId, name).flatMap(_ match {
+      case Some(id: Int) =>
+        db.run(libraries.filter(_.name === name).filter(_.id === id).filter(_.sampleId === sampleId).map(_.tags).update(tags))
+          .map(_ => id)
+      case _ => createLibrary(name, runId, sampleId, tags)
+    })
+  }
+
+  /** Creates a new pipeline, even if it already exist. This may give a database exeption */
+  def forceCreatePipeline(name: String, runId: Int): Future[Int] = {
```

Comment:
This deepens on the database implementation that is used. So this is not in our control, I wise it was.